config/merge/merge_test.go

```diff
@@ -0,0 +1,226 @@
+									},
+								},
+							},
+						},
+						{
+							Node: types.Node{
+								Path: "/bar",
+							},
+							FileEmbedded1: types.FileEmbedded1{
+								Contents: types.FileContents{
+									Compression: util.StrToPtr("gzip"),
+								},
+							},
+						},
+					},
+				},
+			},
+			in2: types.Config{
+				Storage: types.Storage{
+					Disks: []types.Disk{
+						{
+							Device:    "foo",
+							WipeTable: util.BoolToPtr(false),
+							Partitions: []types.Partition{
+								{
+									Number: 1,
+									Label:  util.StrToPtr("labelchanged"),
+								},
+								{
+									Number: 2,
+									Label:  util.StrToPtr("label2"),
+								},
+							},
+						},
+						{
+							Device: "bar",
+						},
+						{
+							Device:    "baz",
+							WipeTable: util.BoolToPtr(true),
+						},
+					},
+					Files: []types.File{
+						{
+							Node: types.Node{
+								Path: "/foo",
+							},
+							FileEmbedded1: types.FileEmbedded1{
+								Append: []types.FileContents{
+									{
+										Source: util.StrToPtr("source1"),
+									},
+									{
+										Source: util.StrToPtr("source2"),
+									},
+								},
+							},
+						},
+					},
+					Directories: []types.Directory{
+						{
+							Node: types.Node{
+								Path: "/bar",
+							},
+						},
+					},
+					Links: []types.Link{
+						{
+							Node: types.Node{
+								Path: "/baz",
+							},
+						},
+					},
+				},
+			},
+			out: types.Config{
+				Storage: types.Storage{
+					Disks: []types.Disk{
+						{
+							Device:    "foo",
+							WipeTable: util.BoolToPtr(false),
+							Partitions: []types.Partition{
+								{
+									Number:   1,
+									Label:    util.StrToPtr("labelchanged"),
+									StartMiB: util.IntToPtr(4),
+								},
+								{
+									Number: 2,
+									Label:  util.StrToPtr("label2"),
+								},
+							},
+						},
+						{
+							Device:    "bar",
+							WipeTable: util.BoolToPtr(true),
+						},
+						{
+							Device:    "baz",
+							WipeTable: util.BoolToPtr(true),
+						},
+					},
+					Files: []types.File{
+						{
+							Node: types.Node{
+								Path: "/foo",
+							},
+							FileEmbedded1: types.FileEmbedded1{
+								Append: []types.FileContents{
+									{
+										Source: util.StrToPtr("source1"),
+									},
+									{
+										Source: util.StrToPtr("source1"),
+									},
+									{
+										Source: util.StrToPtr("source2"),
+									},
+								},
+							},
+						},
+					},
+					Directories: []types.Directory{
+						{
+							Node: types.Node{
+								Path: "/bar",
+							},
+						},
+					},
+					Links: []types.Link{
+						{
+							Node: types.Node{
+								Path: "/baz",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for i, test := range tests {
+		in1v := reflect.ValueOf(test.in1)
+		in2v := reflect.ValueOf(test.in2)
+		out := MergeStruct(in1v, in2v).Interface().(types.Config)
+
+		assert.Equal(t, test.out, out, "#%d bas merge", i)
+	}
```

Comment:
Would be good to have a negative test too to verify the "merging can fail" path. Though I'm not sure if you were planning to add this sort of test at a higher level.